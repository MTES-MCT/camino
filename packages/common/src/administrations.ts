import { Administration, AdministrationId, sortedAdministrations } from './static/administrations.js'
import { CommuneId } from './static/communes.js'
import { DepartementId, Departements, toDepartementId } from './static/departement.js'
import { SecteursMaritimes, getDepartementsBySecteurs } from './static/facades.js'
import { onlyUnique, isNullOrUndefined } from './typescript-tools.js'

// calcule tous les départements d'une étape
const titreEtapeAdministrationsDepartementsBuild = (communes: CommuneId[] | undefined | null, secteursMaritimes: SecteursMaritimes[] | undefined | null): DepartementId[] => {
  if (!communes) {
    throw new Error('les communes ne sont pas chargées')
  }

  const departements = communes.map(id => toDepartementId(id))

  if (secteursMaritimes) {
    departements.push(...getDepartementsBySecteurs(secteursMaritimes))
  }

  return departements.filter(onlyUnique)
}

export const getAdministrationsLocales = (communes: CommuneId[] | undefined | null, secteursMaritimes: SecteursMaritimes[] | undefined | null): AdministrationId[] => {
  const titreDepartementsIds = titreEtapeAdministrationsDepartementsBuild(communes, secteursMaritimes)
  const titreRegionsIds = titreDepartementsIds.map(departementId => Departements[departementId].regionId).filter(onlyUnique)

  // calcule toutes les administrations qui couvrent ces départements et régions
  return sortedAdministrations.reduce((titreEtapeAdministrations: AdministrationId[], administration: Administration) => {
    // si le département ou la région de l'administration ne fait pas partie de ceux de l'étape
    const isAdministrationLocale =
      (administration.departementId && titreDepartementsIds.includes(administration.departementId)) || (administration.regionId && titreRegionsIds.includes(administration.regionId))
    // alors l'administration n'est pas rattachée à l'étape
    if (isNullOrUndefined(isAdministrationLocale) || !isAdministrationLocale) return titreEtapeAdministrations

    titreEtapeAdministrations.push(administration.id)

    return titreEtapeAdministrations
  }, [])
}
