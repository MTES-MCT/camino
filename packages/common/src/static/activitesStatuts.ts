import { z } from 'zod'
import { Couleur } from './couleurs'

const IDS = ['abs', 'enc', 'dep', 'fer'] as const

export const activiteStatutIdValidator = z.enum(IDS)
export type ActivitesStatutId = z.infer<typeof activiteStatutIdValidator>
export const ACTIVITES_STATUTS_IDS = {
  ABSENT: 'abs',
  EN_CONSTRUCTION: 'enc',
  DEPOSE: 'dep',
  CLOTURE: 'fer',
} as const satisfies Record<string, ActivitesStatutId>

export type ActivitesStatut<T = ActivitesStatutId> = {
  id: T
  nom: string
  couleur: Couleur
}
export const ActivitesStatuts: { [key in ActivitesStatutId]: ActivitesStatut<key> } = {
  abs: { id: 'abs', nom: 'absent', couleur: 'error' },
  fer: { id: 'fer', nom: 'cloturé', couleur: 'neutral' },
  dep: { id: 'dep', nom: 'déposé', couleur: 'success' },
  enc: { id: 'enc', nom: 'en construction', couleur: 'warning' },
}

export const activitesStatutsIds = Object.values(ACTIVITES_STATUTS_IDS)
export const activitesStatuts = Object.values(ActivitesStatuts)
