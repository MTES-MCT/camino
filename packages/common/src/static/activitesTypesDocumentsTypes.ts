import { ACTIVITES_TYPES_IDS, ActivitesTypesId } from './activitesTypes.js'
import { DOCUMENTS_TYPES_IDS, DocumentTypeId } from './documentsTypes.js'

export const activitesTypesDocumentsTypes = {
  [ACTIVITES_TYPES_IDS["rapport trimestriel d'exploitation d'or en Guyane"]]: [],
  [ACTIVITES_TYPES_IDS["rapport d'exploitation (permis et concessions M)"]]: [
    {
      documentTypeId: DOCUMENTS_TYPES_IDS.rapportAnnuelDExploitation,
      optionnel: false,
    },
  ],
  [ACTIVITES_TYPES_IDS['rapport d’intensité d’exploration']]: [{ documentTypeId: DOCUMENTS_TYPES_IDS.rapportDIntensiteDExploration, optionnel: true }],
  [ACTIVITES_TYPES_IDS["rapport financier d'exploration"]]: [{ documentTypeId: DOCUMENTS_TYPES_IDS.rapportFinancierDExploration, optionnel: true }],
  [ACTIVITES_TYPES_IDS["rapport environnemental d'exploration"]]: [
    {
      documentTypeId: DOCUMENTS_TYPES_IDS.rapportEnvironnementalDExploration,
      optionnel: true,
    },
  ],
  [ACTIVITES_TYPES_IDS["rapport social et économique d'exploration"]]: [
    {
      documentTypeId: DOCUMENTS_TYPES_IDS.rapportSocialEtEconomiqueDExploration,
      optionnel: true,
    },
  ],
  [ACTIVITES_TYPES_IDS["rapport d'exploitation (permis et concessions W)"]]: [
    {
      documentTypeId: DOCUMENTS_TYPES_IDS.rapportAnnuelDExploitation,
      optionnel: false,
    },
  ],
  [ACTIVITES_TYPES_IDS["rapport d'exploitation (autorisations M)"]]: [{ documentTypeId: DOCUMENTS_TYPES_IDS.rapportAnnuelDExploitation, optionnel: true }],
} as const satisfies { [key in ActivitesTypesId]: Readonly<{ documentTypeId: DocumentTypeId; optionnel: boolean }[]> }
