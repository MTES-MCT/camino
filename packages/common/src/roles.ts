import { AdministrationId } from './administrations'
import { isNotNullNorUndefined } from './typescript-tools'
export const ROLES = [
  'super',
  'admin',
  'editeur',
  'lecteur',
  'entreprise',
  'bureau d’études',
  'defaut'
] as const
export type Role = typeof ROLES[number]
type UserEntreprise = { role: 'entreprise'; administrationId: undefined }
type UserBureaudEtudes = {
  role: 'bureau d’études'
  administrationId: undefined
}
type UserAdmin = { role: 'admin'; administrationId: AdministrationId }
type UserLecteur = { role: 'lecteur'; administrationId: AdministrationId }
type UserEditeur = { role: 'editeur'; administrationId: AdministrationId }
type UserSuper = { role: 'super'; administrationId: undefined }
type UserDefaut = { role: 'defaut'; administrationId: undefined }

type User =
  | { role: Role; administrationId: undefined | null | AdministrationId }
  | undefined
  | null

export const isSuper = (user: User): user is UserSuper =>
  userPermissionCheck(user, 'super')

export const isAdministration = (
  user: User
): user is UserLecteur | UserAdmin | UserEditeur =>
  isAdministrationAdmin(user) ||
  isAdministrationEditeur(user) ||
  isAdministrationLecteur(user)
export const isAdministrationAdmin = (user: User): user is UserAdmin =>
  userPermissionCheck(user, 'admin') &&
  isNotNullNorUndefined(user?.administrationId)
export const isAdministrationEditeur = (user: User): user is UserEditeur =>
  userPermissionCheck(user, 'editeur') &&
  isNotNullNorUndefined(user?.administrationId)
export const isAdministrationLecteur = (user: User): user is UserLecteur =>
  userPermissionCheck(user, 'lecteur') &&
  isNotNullNorUndefined(user?.administrationId)
export const isEntreprise = (user: User): user is UserEntreprise =>
  userPermissionCheck(user, 'entreprise')
export const isBureauDEtudes = (user: User): user is UserBureaudEtudes =>
  userPermissionCheck(user, 'bureau d’études')
export const isDefault = (user: User): user is UserDefaut | undefined =>
  !user || userPermissionCheck(user, 'defaut')

export const isRole = (role: Role | string | undefined | null): role is Role =>
  ROLES.includes(role)

function userPermissionCheck(user: User, role: Role) {
  return user?.role === role
}
