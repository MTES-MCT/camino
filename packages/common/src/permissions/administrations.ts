import { isAdministration, isAdministrationAdmin, isAdministrationEditeur, isSuper, User } from '../roles.js'
import { AdministrationId, Administrations, sortedAdministrations } from '../static/administrations.js'
import { Departements } from '../static/departement.js'

export const canReadActivitesTypesEmails = (user: User, administrationId: AdministrationId) => {
  if (!canReadAdministrations(user)) {
    return false
  }

  if (isSuper(user) || (isAdministration(user) && Administrations[user.administrationId].typeId === 'min')) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    return true
  } else if (isAdministration(user)) {
    const administration = Administrations[user.administrationId]

    const administrationIds: AdministrationId[] = []

    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      administrationIds.push(...sortedAdministrations.filter(({ departementId }) => departementIds.includes(departementId)).map(({ id }) => id))
    }

    administrationIds.push(user.administrationId)

    return administrationIds.includes(administrationId)
  }

  return false
}

export const canReadAdministrations = (user: User) => isSuper(user) || isAdministration(user)

export const canEditEmails = (user: User, administrationId: AdministrationId): boolean => {
  if (isSuper(user) || ((isAdministrationAdmin(user) || isAdministrationEditeur(user)) && Administrations[user.administrationId].typeId === 'min')) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    return true
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    // Membre d'une DREAL/DEAL vis-à-vis de la DREAL elle-même,
    // ou d'un DREAL/DEAL vis-à-vis d'une administration qui dépend d'elles
    // Admin ou éditeur : modifications
    // Admin, éditeur ou lecteur : lecture

    const administration = Administrations[user.administrationId]
    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      const administrationIds = sortedAdministrations.filter(({ departementId }) => departementIds.includes(departementId)).map(({ id }) => id)

      if (administration.typeId === 'dre' || administration.typeId === 'dea') {
        administrationIds.push(user.administrationId)
      }

      return administrationIds.includes(administrationId)
    }
  }

  return false
}
