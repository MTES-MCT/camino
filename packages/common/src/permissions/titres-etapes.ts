import { ETAPES_TYPES, EtapeTypeId } from '../static/etapesTypes.js'
import { TitreTypeId } from '../static/titresTypes.js'
import { DEMARCHES_TYPES_IDS, DemarcheTypeId, isDemarcheTypeWithPhase } from '../static/demarchesTypes.js'
import { isAdministrationAdmin, isAdministrationEditeur, isBureauDEtudes, isEntreprise, isSuper, User } from '../roles.js'
import { EtapeStatutId, ETAPES_STATUTS } from '../static/etapesStatuts.js'
import { TITRES_TYPES_IDS_DEMAT } from './titres.js'
import { AdministrationId } from '../static/administrations.js'
import { isGestionnaire } from '../static/administrationsTitresTypes.js'
import { canAdministrationModifyEtapes } from '../static/administrationsTitresTypesTitresStatuts.js'
import { canAdministrationEtapeTypeId } from '../static/administrationsTitresTypesEtapesTypes.js'

import { TitreStatutId } from '../static/titresStatuts.js'
import { EntrepriseDocument, EntrepriseId } from '../entreprise.js'
import { getSections } from '../static/titresTypes_demarchesTypes_etapesTypes/sections.js'
import { getEntrepriseDocuments } from '../static/titresTypes_demarchesTypes_etapesTypes/entrepriseDocuments.js'
import { SDOMZoneId } from '../static/sdom.js'
import { NonEmptyArray, isNonEmptyArray, isNotNullNorUndefined, isNotNullNorUndefinedNorEmpty, isNullOrUndefined } from '../typescript-tools.js'
import { DocumentType, EntrepriseDocumentTypeId } from '../static/documentsTypes.js'
import { SubstanceLegaleId } from '../static/substancesLegales.js'
import { isDocumentsComplete } from './documents.js'
import { contenuCompleteValidate, sectionsWithValueCompleteValidate } from './sections.js'
import { ElementWithValue, SectionWithValue } from '../sections.js'
import { FeatureMultiPolygon } from '../perimetre.js'
import { EtapeDocument, GetEtapeDocumentsByEtapeId, GetEtapeDocumentsByEtapeIdAslDocument, GetEtapeDocumentsByEtapeIdDaeDocument, needAslAndDae } from '../etape.js'
import { getDocumentsTypes } from './etape-form.js'

export const dureeOptionalCheck = (etapeTypeId: EtapeTypeId, demarcheTypeId: DemarcheTypeId, titreTypeId: TitreTypeId): boolean => {
  if (titreTypeId !== 'axm' && titreTypeId !== 'arm') {
    return true
  }

  if (!isDemarcheTypeWithPhase(demarcheTypeId)) {
    return true
  }

  return etapeTypeId !== 'mfr'
}

export const canEditAmodiataires = (titreTypeId: TitreTypeId, user: User): boolean => {
  // il n’y a pas d’amodiataire sur les ARM et les AXM
  if (titreTypeId === 'arm' || titreTypeId === 'axm') {
    return false
  }

  // seuls les supers et les administrations peuvent éditer les amodiataires
  return isSuper(user) || isAdministrationAdmin(user) || isAdministrationEditeur(user)
}

const demarchesSansDatesNiDureePourLesEtapes = [DEMARCHES_TYPES_IDS.DeplacementDePerimetre, DEMARCHES_TYPES_IDS.Mutation, DEMARCHES_TYPES_IDS.ExtensionDePerimetre] as const
export const canEditDates = (titreTypeId: TitreTypeId, demarcheTypeId: DemarcheTypeId, etapeTypeId: EtapeTypeId, user: User): boolean => {
  // ne peut pas ajouter de dates à la démarche déplacement de périmètre
  if (demarchesSansDatesNiDureePourLesEtapes.includes(demarcheTypeId)) {
    return false
  }

  // peut éditer la date sur les titres autre que ARM et AXM
  if (titreTypeId !== 'arm' && titreTypeId !== 'axm') {
    return true
  }

  // ne peut pas modifier les dates sur une demande d’ARM ou d’AXM car c'est camino qui fait foi
  if (etapeTypeId === ETAPES_TYPES.demande) {
    return false
  }

  // seuls les supers et les administrations peuvent éditer la date de début sur les ARM et les AXM
  return isSuper(user) || isAdministrationAdmin(user) || isAdministrationEditeur(user)
}

export const canEditTitulaires = (titreTypeId: TitreTypeId, user: User): boolean => {
  // peut éditer les titulaires sur les titres autre que ARM et AXM
  if (titreTypeId !== 'arm' && titreTypeId !== 'axm') {
    return true
  }

  // seuls les supers et les administrations peuvent éditer les titulaires sur les ARM et les AXM
  return isSuper(user) || isAdministrationAdmin(user) || isAdministrationEditeur(user)
}

export const canEditDuree = (titreTypeId: TitreTypeId, demarcheTypeId: DemarcheTypeId): boolean => {
  // ne peut pas ajouter de durée à la démarche déplacement de périmètre
  if (demarchesSansDatesNiDureePourLesEtapes.includes(demarcheTypeId)) {
    return false
  }

  // la durée pour les demandes d’ARM est fixée à 4 mois par l’API
  return titreTypeId !== 'arm' || demarcheTypeId !== 'oct'
}

export const canCreateEtape = (
  user: User,
  etapeTypeId: EtapeTypeId,
  etapeStatutId: EtapeStatutId | null,
  titulaires: { id: EntrepriseId }[],
  titresAdministrationsLocales: AdministrationId[],
  demarcheTypeId: DemarcheTypeId,
  titre: { typeId: TitreTypeId; titreStatutId: TitreStatutId }
): boolean => {
  return canCreateOrEditEtape(user, etapeTypeId, etapeStatutId, titulaires, titresAdministrationsLocales, demarcheTypeId, titre, 'creation')
}

export const canEditEtape = (
  user: User,
  etapeTypeId: EtapeTypeId,
  etapeStatutId: EtapeStatutId | null,
  titulaires: { id: EntrepriseId }[],
  titresAdministrationsLocales: AdministrationId[],
  demarcheTypeId: DemarcheTypeId,
  titre: { typeId: TitreTypeId; titreStatutId: TitreStatutId }
): boolean => {
  return canCreateOrEditEtape(user, etapeTypeId, etapeStatutId, titulaires, titresAdministrationsLocales, demarcheTypeId, titre, 'modification')
}

const canCreateOrEditEtape = (
  user: User,
  etapeTypeId: EtapeTypeId,
  etapeStatutId: EtapeStatutId | null,
  titulaires: { id: EntrepriseId }[],
  titresAdministrationsLocales: AdministrationId[],
  demarcheTypeId: DemarcheTypeId,
  titre: { typeId: TitreTypeId; titreStatutId: TitreStatutId },
  permission: 'creation' | 'modification'
): boolean => {
  if (isSuper(user)) {
    return true
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    if (isGestionnaire(user.administrationId, titre.typeId) || titresAdministrationsLocales.includes(user.administrationId)) {
      return canAdministrationModifyEtapes(user.administrationId, titre.typeId, titre.titreStatutId) && canAdministrationEtapeTypeId(user.administrationId, titre.typeId, etapeTypeId, permission)
    }
  } else if (isEntreprise(user) || isBureauDEtudes(user)) {
    return (
      (user.entreprises?.length ?? 0) > 0 &&
      demarcheTypeId === DEMARCHES_TYPES_IDS.Octroi &&
      etapeTypeId === ETAPES_TYPES.demande &&
      etapeStatutId === ETAPES_STATUTS.EN_CONSTRUCTION &&
      TITRES_TYPES_IDS_DEMAT.includes(titre.typeId) &&
      titulaires.some(({ id }) => user.entreprises?.some(entreprise => id === entreprise.id))
    )
  }

  return false
}

type IsEtapeCompleteEtape = {
  typeId: EtapeTypeId
  statutId: EtapeStatutId
  /** 
   @deprecated use sectionsWithValue/
  */
  contenu?: Record<string, Record<string, ElementWithValue['value']>>
  sectionsWithValue?: SectionWithValue[]
  geojson4326Perimetre?: null | FeatureMultiPolygon
  substances?: null | SubstanceLegaleId[]
  duree?: number | null
}

// TODO 2024-04-17 utiliser toutes les stepIsComplete 
export const isEtapeComplete = (
  titreEtape: IsEtapeCompleteEtape,
  titreTypeId: TitreTypeId,
  demarcheTypeId: DemarcheTypeId,
  documents: Pick<EtapeDocument, 'etape_document_type_id'>[],
  entrepriseDocuments: Pick<EntrepriseDocument, 'entreprise_document_type_id'>[],
  sdomZones: SDOMZoneId[] | null | undefined,
  daeDocument: Omit<GetEtapeDocumentsByEtapeIdDaeDocument, 'id'> | null,
  aslDocument: Omit<GetEtapeDocumentsByEtapeIdAslDocument, 'id'> | null,
  user: User
): { valid: true } | { valid: false; errors: NonEmptyArray<string> } => {
  const errors: string[] = []
  const sections = getSections(titreTypeId, demarcheTypeId, titreEtape.typeId)
  // les éléments non optionnel des sections sont renseignés
  const hasAtLeasOneSectionMandatory: boolean = sections.some(section => {
    return section.elements.some(element => (element.optionnel ?? true) === false)
  })
  if (hasAtLeasOneSectionMandatory) {
    if (isNotNullNorUndefined(titreEtape.contenu)) {
      errors.push(...contenuCompleteValidate(sections, titreEtape.contenu))
    } else if (isNotNullNorUndefinedNorEmpty(titreEtape.sectionsWithValue)) {
      errors.push(...sectionsWithValueCompleteValidate(titreEtape.sectionsWithValue))
    } else {
      errors.push('les contenus ne sont pas présents dans l’étape alors que les sections ont des éléments obligatoires')
    }
  }

  if(needAslAndDae({etapeTypeId: titreEtape.typeId, demarcheTypeId, titreTypeId}, titreEtape.statutId, user)){
    if( isNullOrUndefined(daeDocument)){
      errors.push('L’arrêté préfectoral de la mission autorité environnementale est obligatoire')
    }
    if( isNullOrUndefined(aslDocument)){
      errors.push('La lettre de décision du propriétaire du sol est obligatoire')
    }
  }


  const dts: DocumentType[] = getDocumentsTypes({...titreEtape, contenu: titreEtape.contenu ?? {} }, demarcheTypeId, titreTypeId, sdomZones ?? [])

  const documentsErrors = isDocumentsComplete(documents ?? [], dts)
  if (!documentsErrors.valid) {
    errors.push(...documentsErrors.errors)
  }

  // les documents d'entreprise obligatoires sont tous présents
  const entrepriseDocumentsTypes = getEntrepriseDocuments(titreTypeId, demarcheTypeId, titreEtape.typeId)

  const entrepriseDocumentsTypesIds: EntrepriseDocumentTypeId[] = []
  if (entrepriseDocuments.length) {
    for (const entrepriseDocumentType of entrepriseDocuments) {
      if (!entrepriseDocumentsTypes.map(({ id }) => id).includes(entrepriseDocumentType.entreprise_document_type_id)) {
        errors.push(`impossible de lier un document d'entreprise de type ${entrepriseDocumentType.entreprise_document_type_id}`)
      }
      entrepriseDocumentsTypesIds.push(entrepriseDocumentType.entreprise_document_type_id)
    }
  }
  entrepriseDocumentsTypes
    .filter(({ optionnel }) => !optionnel)
    .forEach(jt => {
      if (!entrepriseDocumentsTypesIds.includes(jt.id)) {
        errors.push(`le document d'entreprise « ${jt.nom} » obligatoire est manquant`)
      }
    })

  // Si c’est une demande d’AEX ou d’ARM, certaines informations sont obligatoires
  if (titreEtape.typeId === 'mfr' && ['arm', 'axm'].includes(titreTypeId)) {
    // le périmètre doit être défini
    if (isNullOrUndefined(titreEtape.geojson4326Perimetre)) {
      errors.push('le périmètre doit être renseigné')
    }

    // il doit exister au moins une substance
    if (!titreEtape.substances || !titreEtape.substances.length || !titreEtape.substances.some(substanceId => !!substanceId)) {
      errors.push('au moins une substance doit être renseignée')
    }
  }

  if ((isNullOrUndefined(titreEtape.duree) || titreEtape.duree === 0) && !dureeOptionalCheck(titreEtape.typeId, demarcheTypeId, titreTypeId)) {
    errors.push('la durée doit être renseignée')
  }

  if (isNonEmptyArray(errors)) {
    return { valid: false, errors }
  }

  return { valid: true }
}

export const isEtapeDeposable = (
  user: User,
  titre: {
    typeId: TitreTypeId
    titreStatutId: TitreStatutId
    titulaires: { id: EntrepriseId }[]
    administrationsLocales: AdministrationId[]
  },
  demarcheTypeId: DemarcheTypeId,
  titreEtape: IsEtapeCompleteEtape,
  etapeDocuments: Pick<EtapeDocument, 'etape_document_type_id'>[],
  entrepriseDocuments: Pick<EntrepriseDocument, 'entreprise_document_type_id'>[],
  sdomZones: SDOMZoneId[] | null | undefined,
  daeDocument: GetEtapeDocumentsByEtapeId['dae'],
  aslDocument: GetEtapeDocumentsByEtapeId['asl'],
): boolean => {
  if (titreEtape.typeId === ETAPES_TYPES.demande && titreEtape.statutId === ETAPES_STATUTS.EN_CONSTRUCTION) {
    if (
      canCreateOrEditEtape(
        user,
        titreEtape.typeId,
        titreEtape.statutId,
        titre.titulaires,
        titre.administrationsLocales,
        demarcheTypeId,
        { typeId: titre.typeId, titreStatutId: titre.titreStatutId },
        'modification'
      )
    ) {
      const complete = isEtapeComplete(titreEtape, titre.typeId, demarcheTypeId, etapeDocuments, entrepriseDocuments, sdomZones, daeDocument, aslDocument, user)
      if (!complete.valid) {
        console.warn(complete.errors)

        return false
      }

      return true
    }
  }

  return false
}

export const canDeleteEtapeDocument = (etapeStatutId: EtapeStatutId | null): boolean => etapeStatutId === null || etapeStatutId === ETAPES_STATUTS.EN_CONSTRUCTION
