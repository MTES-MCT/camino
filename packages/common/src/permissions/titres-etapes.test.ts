import { TitresTypesIds, TitreTypeId } from '../static/titresTypes'
import { EtapeTypeId } from '../static/etapesTypes'
import { DemarcheTypeId } from '../static/demarchesTypes'
import { canEditAmodiataires, canEditDates, canEditDuree, canEditTitulaires, dureeOptionalCheck } from './titres-etapes'
import { User } from '../roles'
import { ADMINISTRATION_IDS } from '../static/administrations'

test.each<{ etapeTypeId: EtapeTypeId; demarcheTypeId: DemarcheTypeId; titreTypeId: TitreTypeId; optional: boolean }>([
  { etapeTypeId: 'mfr', demarcheTypeId: 'oct', titreTypeId: 'arm', optional: false },
  { etapeTypeId: 'mfr', demarcheTypeId: 'oct', titreTypeId: 'axm', optional: false },
  { etapeTypeId: 'dex', demarcheTypeId: 'oct', titreTypeId: 'axm', optional: true },
  { etapeTypeId: 'mfr', demarcheTypeId: 'oct', titreTypeId: 'prm', optional: true },
  { etapeTypeId: 'mfr', demarcheTypeId: 'dep', titreTypeId: 'arm', optional: true }
])('dureeOptionalCheck $etapeTypeId | $demarcheTypeId | $titreTypeId | $optional', ({ etapeTypeId, demarcheTypeId, titreTypeId, optional }) => {
  expect(dureeOptionalCheck(etapeTypeId, demarcheTypeId, titreTypeId)).toEqual(optional)
})

test.each<TitreTypeId>(TitresTypesIds)('canEditDuree %p', titreTypeId => expect(canEditDuree(titreTypeId)).toMatchSnapshot())

test.each<{ titreTypeId: TitreTypeId; etapeTypeId: EtapeTypeId; user: User; canEdit: boolean }>([
  { titreTypeId: 'arm', etapeTypeId: 'mfr', user: { role: 'super', administrationId: undefined }, canEdit: false },
  { titreTypeId: 'arm', etapeTypeId: 'dpu', user: { role: 'super', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'axm', etapeTypeId: 'mfr', user: { role: 'super', administrationId: undefined }, canEdit: false },
  { titreTypeId: 'prm', etapeTypeId: 'mfr', user: { role: 'super', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'prm', etapeTypeId: 'mfr', user: { role: 'admin', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'prm', etapeTypeId: 'mfr', user: { role: 'lecteur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true }
])('canEditDate $titreTypeId | $etapeTypeId | $user | $canEdit', ({ titreTypeId, etapeTypeId, user, canEdit }) => {
  expect(canEditDates(titreTypeId, etapeTypeId, user)).toEqual(canEdit)
})

test.each<{ titreTypeId: TitreTypeId; user: User; canEdit: boolean }>([
  { titreTypeId: 'arm', user: { role: 'super', administrationId: undefined }, canEdit: false },
  { titreTypeId: 'axm', user: { role: 'super', administrationId: undefined }, canEdit: false },
  { titreTypeId: 'prm', user: { role: 'super', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'admin', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'editeur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'lecteur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: false },
  { titreTypeId: 'prm', user: { role: 'defaut', administrationId: undefined }, canEdit: false }
])('canEditAmodiataires $titreTypeId | $user | $canEdit', ({ titreTypeId, user, canEdit }) => {
  expect(canEditAmodiataires(titreTypeId, user)).toEqual(canEdit)
})

test.each<{ titreTypeId: TitreTypeId; user: User; canEdit: boolean }>([
  { titreTypeId: 'prm', user: { role: 'super', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'admin', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'editeur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'lecteur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'prm', user: { role: 'defaut', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'axm', user: { role: 'super', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'axm', user: { role: 'admin', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'axm', user: { role: 'editeur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'axm', user: { role: 'lecteur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: false },
  { titreTypeId: 'axm', user: { role: 'defaut', administrationId: undefined }, canEdit: false },
  { titreTypeId: 'arm', user: { role: 'super', administrationId: undefined }, canEdit: true },
  { titreTypeId: 'arm', user: { role: 'admin', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'arm', user: { role: 'editeur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: true },
  { titreTypeId: 'arm', user: { role: 'lecteur', administrationId: ADMINISTRATION_IDS.BRGM }, canEdit: false },
  { titreTypeId: 'arm', user: { role: 'defaut', administrationId: undefined }, canEdit: false }
])('canEditTitulaires $titreTypeId | $user | $canEdit', ({ titreTypeId, user, canEdit }) => {
  expect(canEditTitulaires(titreTypeId, user)).toEqual(canEdit)
})
