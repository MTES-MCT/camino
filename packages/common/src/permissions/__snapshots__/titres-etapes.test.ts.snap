// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`canEditDuree "apc" 1`] = `true`;

exports[`canEditDuree "aph" 1`] = `true`;

exports[`canEditDuree "apm" 1`] = `true`;

exports[`canEditDuree "apw" 1`] = `true`;

exports[`canEditDuree "arc" 1`] = `true`;

exports[`canEditDuree "arg" 1`] = `true`;

exports[`canEditDuree "arm" 1`] = `false`;

exports[`canEditDuree "axm" 1`] = `true`;

exports[`canEditDuree "cxf" 1`] = `true`;

exports[`canEditDuree "cxg" 1`] = `true`;

exports[`canEditDuree "cxh" 1`] = `true`;

exports[`canEditDuree "cxi" 1`] = `true`;

exports[`canEditDuree "cxm" 1`] = `true`;

exports[`canEditDuree "cxr" 1`] = `true`;

exports[`canEditDuree "cxs" 1`] = `true`;

exports[`canEditDuree "cxw" 1`] = `true`;

exports[`canEditDuree "inh" 1`] = `true`;

exports[`canEditDuree "ini" 1`] = `true`;

exports[`canEditDuree "inm" 1`] = `true`;

exports[`canEditDuree "inr" 1`] = `true`;

exports[`canEditDuree "pcc" 1`] = `true`;

exports[`canEditDuree "prf" 1`] = `true`;

exports[`canEditDuree "prg" 1`] = `true`;

exports[`canEditDuree "prh" 1`] = `true`;

exports[`canEditDuree "pri" 1`] = `true`;

exports[`canEditDuree "prm" 1`] = `true`;

exports[`canEditDuree "prr" 1`] = `true`;

exports[`canEditDuree "prs" 1`] = `true`;

exports[`canEditDuree "prw" 1`] = `true`;

exports[`canEditDuree "pxf" 1`] = `true`;

exports[`canEditDuree "pxg" 1`] = `true`;

exports[`canEditDuree "pxh" 1`] = `true`;

exports[`canEditDuree "pxi" 1`] = `true`;

exports[`canEditDuree "pxm" 1`] = `true`;

exports[`canEditDuree "pxr" 1`] = `true`;

exports[`canEditDuree "pxw" 1`] = `true`;
